### PR TITLE
Fix clang 12 warnings

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This fixes most of the warnings I encountered when building with qmcpack with clang 12. I'm hoping that adding the virtual destructor suggested by `-Wdelete-non-abstract-non-virtual-dtor` might eliminate some of the indirect memory leaks reported by address sanitizer. Consistently labels structs and classes as recommended by `-Wmismatched-tags` sounds smart and adding the missing override keyword is good practice.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- n/a. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- n/a. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCWaveFunctions/BsplineFactory/HybridRepCenterOrbitals.h [hunk 1]
```diff
@@ -708,7 +708,7 @@ class HybridRepCenterOrbitals
   }
 
   template<class BSPLINESPO>
-  friend class HybridRepSetReader;
+  friend struct HybridRepSetReader;
```
Comment (ye-luo) [inline src/QMCWaveFunctions/BsplineFactory/HybridRepCenterOrbitals.h:711]: We really prefer class. So even just fix as `class{ public:` is better than `struct`

Comment (quantumsteve) [reply src/QMCWaveFunctions/BsplineFactory/HybridRepCenterOrbitals.h]: updated

Comment (prckent) [reply src/QMCWaveFunctions/BsplineFactory/HybridRepCenterOrbitals.h]: Thanks Steve. There are plenty of structs that should be classes.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**